### ISSUE TITLE
Code generation experience needs improvements

### ISSUE THREAD

**Issue by szotp** [flutter/flutter#63323]
## Use case

Code generation has became a big part of flutter development, I'm currently using it for:

1. Generating i18n strings.
2. Generating asset paths.
3. Generating json parsing.
4. Generating moor database handling.

Unfortunately, the experience related to code generation is not as good as it should be:

1. Code generated files are checked by analyzer for styling problems.
2. I need to setup build_runner for every project and tell people how to use it.
3. I need to remember to run build_runner.
4. Build runner takes many seconds to build. 
5. During build process, generated code gets deleted, and analyzer errors are being displayed.
6. Using a code generator requires boilerplate setup for each class:
a. Adding part directive.
b. Adding factory methods that delegate constructors to the generated class.
7. Generated files pollute my project's workspace, project wide text search returns results from generated code.
8. Analysis exclude does not work properly: https://github.com/dart-lang/sdk/issues/25551, and if it worked properly, then I would not see errors.

I am aware that some of these problems have partial solutions, such as using snippets, configuring IDE to run build_runner task, and so on. I just don't think it's good enough if everyone who uses code generation needs to do it on their own. 

## Proposal
Code generation should be completely seamless. Below are my suggestions that would achieve this. I realize that some of them may be infeasible.

1. Code generation should be integral part of build process. IDE must generate code automatically and keep it up to date.
2. All generated code should be send to build cache, unless I'm publishing a package.
3. Analyzer should not show any errors if code is not generated.
4. Analyzer should show generation errors. The errors should be as descriptive as possible.
5. Code generation should be fast.
6. It should be possible to code generate static methods and costructors for the annotated class.
7. 'part' directive should not be needed. Adding annotation should be enough.
8. If there are any errors in generated code, they should be treated as code generation errors. Styling issues should be ignored completely.

## Example

Using json_serializable as example, here is current state of things:
```yaml
dependencies:
  json_annotation: ^3.0.1

dev_dependencies:
  build_runner: ^1.8.1
  json_serializable: ^3.3.0
```
```dart
import 'package:json_annotation/json_annotation.dart';

part 'example.g.dart';

@JsonSerializable(nullable: false)
class Person {
  final String firstName;
  final String lastName;
  final DateTime dateOfBirth;
  Person({this.firstName, this.lastName, this.dateOfBirth});
  factory Person.fromJson(Map<String, dynamic> json) => _$PersonFromJson(json);
  Map<String, dynamic> toJson() => _$PersonToJson(this);
}
```

And here is what I would like to write:
```yaml
builders:
  json_serializable: ^3.3.0
```
```dart
import 'package:json_annotation/json_annotation.dart';

@JsonSerializable(nullable: false)
class Person {
  final String firstName;
  final String lastName;
  final DateTime dateOfBirth;
}
```

## Benefits

1. Parsing jsons will become pleasant experience.
2. People will start writing smaller code generators that were too infeasible/cumbersome before.
3. People will start experimenting with different approaches to flutter. See: https://github.com/flutter/flutter/issues/51752#issuecomment-671028795
4. Inexperienced flutter devs will be able to use it.

**Comment by Hixie** [comment]
cc @jonahwilliams @leafpetersen

**Comment by Hixie** [comment]
For the first item (Code generated files are checked by analyzer for styling problems), I would be worried about excluding some lints that are not about formatting but about actual potential logic errors. (Also, if I ever generate code, I would want it to also follow formatting rules, because that makes it much easier to debug.) Does `ignore_in_file` not solve this use case?

**Comment by jonahwilliams** [comment]
@jakemac53

**Comment by szotp** [comment]
> Does `ignore_in_file` not solve this use case?

Not really.
1. There is no consensus on styling. So package authors need to care enough to install maximally pedantic rules and either use ignore_in_file or follow them.
2. It's nice if generated code follows those guidelines, but it's not super important. For the same reason I can live with third party dependencies that look messy, as long as they work correctly. In this regard, generated code is conceptually closer to package that generated it than my own code.
3. If there is logic error that breaks the compilaton, then the package is clearly broken. So one clear error should be enough: "package X generated invalid code. Contact the author."

**Comment by maks** [comment]
We basically actually (temporarily) had something like what @szotp you ask for until #61508 deprecated and removed that functionality. 

But despite the points in the comment of that issue, I don't think that asking users to just do `pub run build_runner` is good enough, again given how pervasive code generation is in Flutter code because of the restriction on runtime reflection (which I actually think *is* a good thing!) having a much more seamless code generation tooling UX would really help. Coming from a Android background, I contrast with how Android Studio and its tooling provides a much more seamless experience code gen UX.

**Comment by szotp** [comment]
Yes, I was sad to see `flutter generate` go, it seemed like a good start. I'm hoping that this ticket will make flutter devs reconsider their stance. For me, after nullables, it's one of the biggest hurdles of Flutter ecosystem. In Swift* I could just slap Equatable / Codable onto class and compiler would generate implementations. 

A good example of code generation handling would be what they did in C#, I would love to see something like this in Dart:
https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/

* note that Swift does not have what I'm proposing, because the code generation is hardcoded into the compiler and is not extensible

**Comment by maks** [comment]
I should also add that from the point of view as a *writer* of a package doing code-gen I found working with the build package a good experience and much less intimidating then trying to do so with for example Kotlin/Java code gen.

Its just on the code gen user/consumer side that I find the experience less than ideal. I'm pretty much in the same camp: I'd like to just add a builder dependency and have the flutter tooling do the rest for me. 

To be honest I don't really see the reasoning behind any of the reasons given for the `flutter generate` functionality being abandoned? 

From a outsiders point of view, what I saw was: the functionality broke soon after it was introduced, [a bug was filed](https://github.com/flutter/flutter/issues/42003) while the bug was only in `dev`, there was silence (apart from my and a couple of others commenting on aspects of what may be the cause of the issue) on it for months as the breakage slowly made its way down into stable, more silence and then all of a sudden the functionality was removed :disappointed: .

If the Flutter team would prefer not to commit to building any specific functionality around this I can understand that view point, but it would then be good if we could have some sort of way of integrating into the flutter tooling to extend it for ourselves. I had thought that #25377 covered that request, but I've been told that's not the case so do we need to request that as part of this issue or should a new issue be raised to request that?

**Comment by jakemac53** [comment]
The issues stated here are definitely valid, but there are a lot of roadblocks to trying to resolve them.

Note that building support directly into the compiler here would be the most performant and ergonomic thing to do - but that has a _lot_ of implications:

- Currently the `build` package relies on `analyzer` apis, but the analyzer does not exist during compilation. Instead the `kernel` package is used, and so migrating to that would break all existing code generators. 
- Currently the `kernel` package does not guarantee any amount of stability, and is no longer updated on pub at all (it was only published for a while because analyzer depended on it, but was never intended for public use). Maintaining internal compiler details as a public package like this incurs significant technical debt on the kernel team, and could potentially slow down the evolution of the language as a whole.
- We have no way of dynamically loading code in Dart, which would be the most intuitive way of loading these code generators. We could take the `build_runner` approach here which generates a custom build script before even starting compilation, but that comes at a significant cost. 

Given how breaking any meaningful changes here would be, we would want to essentially start from square one and investigate general meta programming and/or static reflection techniques, imo.

Barring compiler integration, we could make some smaller changes to alleviate some issues (integrate flutter more directly with build_runner, support `external` declarations which can be filled in by builders to remove references to generated symbols). But, that ultimately won't be able to provide the experience you desire, imo.

Note also that the existing code generation is a lot more powerful generally than what C# has built in (or any language I am aware of for that matter), because it can do a lot more than Dart -> Dart source code generation. It supports arbitrary code generation, and is really an entire build system separate from the compiler. That is a lot more powerful than anything that would be directly integrated into the compilers, but it is also a lot heavier weight than what most people need.

**Comment by maks** [comment]
Thanks @jakemac53 those are very good points.

Personally I'm not in favour of the support directly in the compiler as for me, through the lens of a Android developer, a very important use case is code generation from "resources", eg what [r_flutter](https://pub.dev/packages/r_flutter) or my own simple [string_res](https://pub.dev/packages/string_res) packages do.

 >Barring compiler integration, we could make some smaller changes to alleviate some issues (integrate flutter more directly with build_runner, support external declarations which can be filled in by builders to remove references to generated symbols). 

I would like that as a starting point. While it might not provide 100% of the experience, I think it would go a very long way towards   it and would be most welcome. 

@jakemac53  from your point of view, what is the way forward on getting more direct flutter integration for build_runner?

**Comment by jakemac53** [comment]
> @jakemac53 from your point of view, what is the way forward on getting more direct flutter integration for build_runner?

The biggest thing would be convincing the flutter developers that it is worth while to do. We would be happy to work on any blockers from our side (I am one of the owners of build_runner, but I am not on the flutter team).

There was at one point some integration already, but not really documented I don't think.

**Comment by maks** [comment]
@jakemac53 ah sorry about that, I did think that you wer on the flutter team. The integration that was there was [documented](https://github.com/flutter/flutter/wiki/Code-generation-in-Flutter/c19420f8474f611cd1f6c4d499c1111d3f407f41) but that's been removed now as per #61508  which deprecated and removed that support.

One of the reasons given for that removal was the "large API surface area" of build_runner but that's something I don't really understand, in using build_runner programmatically would there not be just a single API entry point to get the equivalent of `pub run build_runner` ?

**Comment by jakemac53** [comment]
> One of the reasons given for that removal was the "large API surface area" of build_runner but that's something I don't really understand, in using build_runner programmatically would there not be just a single API entry point to get the equivalent of `pub run build_runner` ?

Alas it was a bit more complicated than that - build_runner really needs to stay alive as a long lived process to be efficient, but flutter doesn't want it to arbitrarily do rebuilds (it wants it to control when those builds happen - after a user initiates a hot restart or reload).

I am not totally sure but I believe this functionality was implemented in flutter by using the `build_daemon` package which is an imperative api that talks to a build_runner process and controls builds. It is definitely a bit more work to set up, and was a new and not heavily used api so it has had some breaking changes as it has evolved.

I don't know if the issues were with the build_daemon api or what here, as there aren't any outstanding issues filed on our repo that I am aware of... my sense is this was more of a general simplification effort? Note that there were some outstanding issues like build to "cache" builders not working correctly with flutter (although this could easily be solved).

**Comment by szotp** [comment]
I have submitted a simpler feature request that would help quite a bit: https://github.com/dart-lang/sdk/issues/43616

**Comment by marcellodesales** [comment]
# Same discussion on Freezed classes

* Using https://pub.dev/packages/freezed for better code patterns
* I have a team pushing the `"freezed"` code to git
* I'd like to have an integrated `flutter run` where one of the stages of the `run` command is the actual `processResources`, similar to Gradle builds for Java/Android code, which optionally generates any source-code that registers to a stage prior to the `build` where source-code is generated and placed under the `build` directory as additional sources to the project

**Comment by gnprice** [comment]
Adding a cross-reference because I don't see it already mentioned in this thread:

The Dart team have an in-progress proposal for adding static metaprogramming to Dart in the form of macros. Background and proposed spec:
https://github.com/dart-lang/language/issues/1482
https://github.com/dart-lang/language/blob/master/working/macros/motivation.md
https://github.com/dart-lang/language/blob/master/working/macros/feature-specification.md

If I understand correctly, that can be seen as the Dart team's current thinking on the best strategy for meeting this need (as well as other needs, like automating the boilerplate of using Flutter `StatefulWidget`.)

Macros would be an exciting feature! After the Dart language folks finish their work on the exciting features like records and pattern matching that are planned for Dart 3.0, I hope we'll see macros move forward sometime soon.

**Comment by szotp** [comment]
[Macros got cancelled](https://medium.com/dartlang/an-update-on-dart-macros-data-serialization-06d3037d4f12) so I feel like it's time to reconsider this proposal.

**Comment by davidmorgan** [comment]
Thanks @szotp. 

Exactly as you say, since we're no longer doing macros, all these issues are higher priority :)

For item 4 on your list, I will be working on build_runner performance here: https://github.com/dart-lang/build/issues/3800

We are still working on augmentations, which will address 6b; I'd love to tackle 6a, too.

I'll let others more familiar with the UX questions comment on those :)

**Comment by Mikkelet** [comment]
If we could get rid of the `part` statement for generated files, it would be very helpful.

If not, I would love it if we could ignore the risk of filename conflicts and just put all generated files into a single folder with no subfolders. This would personally help my projects

**Comment by matanlurey** [comment]
Removing `team-tool` as it's unlikely the `flutter CLI` specifically is going to be the source of change.

**Comment by tirth-patel-nc** [comment]
Added `team-framework` & `customer: crowd`

**Comment by davidmorgan** [comment]
These are all still great suggestions :)

`build_runner` 2.5.0 released a few weeks back, now there is `build_runner` 2.6.0 coming soon for the [element 2](https://github.com/dart-lang/build/discussions/4070) migration. Then I'll be unblocked to work on smaller improvements including performance and quality of life improvements--and to start planning for bigger things.

I think broadly we understand the problems at this point and are working on them.

There is now

https://github.com/dart-lang/build/discussions

for general discussion around `build_runner` and related topics, any input and feedback welcome. I'll aim to publish updates there about what's planned / happening.

Thanks!
